Update Erlang syntax for recent language changes

jchristgit opened the pull request:
Erlang recently added the `-moduledoc` attribute as well as triple quoted strings and the `~` prefix for binary strings, see [1].

I think the `-moduledoc` and `-doc` attributes with a comment are not fully highlighted properly, because the attributes themselves are then "highlighted" as comments. I tried to fix this by adding another `syn match` that I thought should attribute these on its own, but then the remaining highlighting turns back into a regular string. Would be happy for pointers here.

[1]: https://www.erlang.org/doc/system/documentation.html

hcs42 commented:
Thank you for the commit.

How does `test_syntax.erl` look around line 268 when you set `let g:erlang_use_markdown_for_docs = 1`? For me, it continues to highlight the lines as docstrings. If it works fine for you: which version of markdown.vim do you use? Is it the one built into Vim, or a plugin? It it's built in, which Vim version or commit do you use?

![image](https://github.com/user-attachments/assets/3ff5fd71-954a-4745-b1ed-6eba0da0ecae)

hcs42 commented:
I also played around with adding support for Erlang 27 docstring when working on an Erlang 27 project. It works for my use cases, but it's not perfect, see below. Afterwards I got back to projects with older Erlang versions and did not want to push or merge my unfinished code.

Now however, I cleaned up my code in [this commit](https://github.com/vim-erlang/vim-erlang-runtime/commit/05412514b2a5de34453a9f4e1901193bd6c90c85) and pushed it to my [erlang_27_docstrings](https://github.com/hcs42/vim-erlang-runtime/tree/erlang_27_docstrings) branch. It's still incomplete. My suggestion is you can move its basic logic into your commit, and merge it with the missing features that are present in your commit. Technically, it might be easiest if you use my code as the basis and add the missing stuff on the top of that. You can use your commit as the basis so that you remain the author in Git. In the commit message, you can mention both of us as the authors. In the end, we would have one commit that contains all new code.

Here are my remarks and pointers.

# Your commit

* I will soon send my review comments via GitHub' review functionality.

# My commit: Goals

* (1) My approach was to handle the `-moduledoc` attribute and the triple quoted strings separately.
    - Pro: It was easier to implement this way.
    - Con: This way all triple quoted strings are highlighted as markdown (even those inside functions). This is one of the reasons why I view my commit as unfinished.
    - It would be nice if you could take my logic of having separate syntax items for the `-moduledoc` attribute and for the string part of the docstring, and add your logic of combining them into a new syntax element. This way we could avoid highlighting all triple quoted strings as Markdown. You might find either the `contains`/`contained` arguments or the `\zs` pattern useful for this.
* (2) My approach towards highlighting (that is, the desired colors) was the following:
    - Highlight the `-moduledoc` characters with the same color as `-module`. This is more consistent and less surprising (at least to me) that highlighting `-moduledoc` differently.
    - Highlight the `"""` characters with the same color as `"`. Same reasoning.
    - Highlight the `.` characters at the and with the same color as `.` normally (no highlight). Same reasoning.
    - Highlight the inside of the docstring according to `g:erlang_use_markdown_for_docs` and `g:erlang_docstring_default_highlight`.
* (3) Should we enable Markdown by default?
    - `g:erlang_use_markdown_for_docs = 1`: My preference is enabling the Markdown highlighting because:
        - I think most people would want to use it.
        - If we disable it by default, it won't occur to almost anyone that this is something that they can enable.
    - If we enable it by default, and some people want to disable it, it will hopefully occur to them that they can do that by configuring the plugin.
    - See Screenshot 1-2-3 versus screenshot 4.
* (4) How should we highlight the plain parts of the Markdown text?
    - It's not easy to make a decision about the default value of `g:erlang_docstring_default_highlight'`. I see three sensible default values.
    - (a) `g:erlang_docstring_default_highlight = ''`:
        - See Screenshot 1.
        - My preference is highlighting all characters in the docstring the same way as in Markdown, including the plain text parts.
        - For example, with the default colors on a white background this means that plain Markdown text inside docstrings will be black, not pink (string) or blue (comments).
        - Otherwise the formatted elements will look somewhat weird:
            - Screenshot 2: If the plain text is pink, words in italic, bold and inline code will still be black, even though the text around them is pink. On the other hand, Markdown headers and link will be just as pink as the plain Markdown text.
            - Screenshot 3: If the plain text is blue, words in italic, bold and inline code will still be black, even though the text around them is blue. On the other hand, Markdown comments will be just as blue as the plain Markdown text.
        - The drawback is that it might be difficult to tell the documentation and the code apart just by looking at it.
    - (b) `g:erlang_docstring_default_highlight = 'String`:
        - See Screenshot 2.
        - My secondary preference is highlighting the plain parts of docstrings as strings.
        - In Vim, docstrings that are syntactically strings are usually highlighted as strings (for example, in Python and Julia). The Erlang 27 docstrings are also syntactically strings.
        - In Vim, docstrings that are syntactically comments are usually highlighted as comments (for example, in Java and C#). The Edoc documentation is also syntactically comments.
        - Other tools will mostly highlight docstrings as strings, so with option (b) the developers need the least amount of cognitive effort to switch between Vim and for example GitHub.
    - (c) `g:erlang_docstring_default_highlight = 'Comment`:
        - See Screenshot 3.
        - I prefer this option the least: highlighting the plain parts of docstrings as strings.
        - I see value in easily seeing which lines are documentation (proper Markdown, goes into `ex_doc`, etc) and which lines are just plain comments.
        - However, I can see that if people got used to having documentation highlighted as comments, they want to continue seeing it that way. With the default setup, this would mean that the function bodies would be separated by blocks of blue documentation, just as before.
    - Maybe we should survey our fellow developers who use Vim, and ask them which highlight they would want to see in Erlang 27 docstrings?

# My commit: Implementation

* (5) I prefer setting up variables such as `s:old_style` for configuration, and then using those variables. I moved this variable to the top of the script and added the new configuration parameters below.
* (6) The `syn region erlangStringTripleQuoted matchgroup=String start=/"""/ end=/\%(^\s*\)\@<="""/ contains=@markdown keepend` line does the heavy lifting.
    - `matchgroup=String`: This highlights the `"""` characters as a string.
    - `/\%(^\s*\)\@<="""/`: According to [EEP 64: Triple-Quoted Strings](https://www.erlang.org/eeps/eep-0064#triple-quoted-string-end), this is how triple-quoted strings end:
      > A triple-quoted string ends with newline followed by optional white-space and then """. This completes the scanner token.
    - Without `keepend`, the highlight will be incorrect for incomplete Markdown elements. For example, in the following line, Vim will highlight everything from the asterisk as italic, because it tries to find where the italic stops:

        ```
        -moduledoc """
        *italic
        """.
        ```

      Even if the Markdown text is correct, cases like this happen because Markdown syntax highlight is difficult (maybe impossible) to get right with Vim's syntax highlighting tools. So we must always stop the docstring block at the closing `"""`, no matter what the Markdown plugin thinks. `keepend` achieves this.
* (7) My implementation does not handle one-liner docstrings. It would be nice if you could add this, based on your experience when experimenting with your commit.

# Outside the scope of this commit, but worth mentioning

* (8) The `syntax iskeyword @,48-57,192-255,$,_` line was the result of some debugging. The top of the `erlang.vim` executes `setlocal iskeyword+=$,@-@`. Notice that one is `syntax keyword` while the other is `setlocal iskeyword`. Maybe this could be improved a bit. One would need to look at how other filetype plugins do this. For example, maybe `ftplugin/erlang.vim` should call `setlocal iskeyword=@,48-57,192-255,$,_`, and `syntax/erlang.vim` should call `syntax iskeyword @,48-57,192-255,$,_` regardless of using Markdown or not. But it works now, so you don't have to worry about it; I just wanted to write this down.
* (9) `:help erlang` currently describes options in `syntax/erlang.vim` that we removed in 2013. After we complete this commit and merge this PR, and I create a corresponding PR in the main Vim repository, I should update this help text.

# Screenshots

## Screenshot 1: Default

```
let g:erlang_use_markdown_for_docs = 1
let g:erlang_docstring_default_highlight = ''
```

![1](https://github.com/user-attachments/assets/82494c1e-3103-43c3-b313-ed81a04c0ca0)

## Screenshot 2: String for plain text

```
let g:erlang_use_markdown_for_docs = 1
let g:erlang_docstring_default_highlight = 'String'
```

![2](https://github.com/user-attachments/assets/6e27a16a-26b9-40e0-8e83-4be9086e9070)

## Screenshot 3: Comment for plain text

```
let g:erlang_use_markdown_for_docs = 1
let g:erlang_docstring_default_highlight = 'Comment'
```

![3](https://github.com/user-attachments/assets/ef7037f8-05e7-4f14-95ea-f1bd4f458d1e)

## Screenshot 4: No Markdown

```
let g:erlang_use_markdown_for_docs = 0
let g:erlang_docstring_default_highlight = 'String'
```

![4](https://github.com/user-attachments/assets/7509d713-6218-446e-a4d9-89c9ed67deb2)

hcs42 commented:
> But I agree we should poll the community. Perhaps an Erlang Forums post might be a good idea?

The Erlang Forum sounds like a good idea. But before polling, we can discuss it a bit more (I would like to understand the situation in other languages/editors better).

Do you use an Elixir plugin to achieve the Markdown syntax higlighting? My Vim simple higlights `@moduledoc` and `@doc` in Elixir as a string (no Markdown). My Vim version is more than a year old, so if you have the Elixir Markdown support built-in, I might need to build a newer version.

I will probably have a busy week, so I might not be able to look at the new changes until the weekend.

hcs42 commented on syntax/erlang.vim:
```diff
@@ -125,11 +125,19 @@ syn match erlangNumberFloat   '\<\d\+\.\d\+\%([eE][+-]\=\d\+\)\=\>'
+syn region erlangStringTripleQuoted matchgroup=String start=/"""/ end=/\%(^\s*\)\@<="""/ keepend
+
+" Documentation
+syn cluster erlangDocStringContains contains=erlangInnerDocAttribute,erlangDocStringDelimiter
```
I suggest naming this `erlangDocStringCluster` (since it's a `syntax cluster`).

Also, I suggest moving it below the four `erlangDocString` lines. That way we keep a top-down order and it's easier to understand this block (`erlangDocString` contains `erlangDocStringCluster` , which contains `erlangDocStringDelimiter`).

hcs42 commented:
> After a lot more local testing, I think the latest commit should just work.

Thank you, I'm happy with it! I have a small stylistic suggestion, which I sent as a review comment.

Please squash the two commits together (yours and mine), and keep the commits message of yours.

I am OK with setting `g:erlang_docstring_default_highlight` to `Comment`, since vim-elixir does that too. So please either update the commit accordingly, or poll people on Erlang Forums.

> This delay is large enough (`time2` is with markdown enabled) that I think we should not enable it by default. What do you think?

What did you measure exactly? Can you measure the same thing with `g:erlang_use_markdown_for_docs`?

I was playing around with trying to measure the difference, and found the following. I used the `lib/kernel/src/socket.erl` from <https://github.com/erlang/otp> because it's quite long (6918 lines).

* When opening the `lib/kernel/src/socket.erl` module, it is done quickly regardless of `g:erlang_use_markdown_for_docs`.
* When I hit `G` to jump to the end:
    - `g:erlang_use_markdown_for_docs = 0` → it jumps almost instantly.
    - `g:erlang_use_markdown_for_docs = 1` → it takes around 1 second to jump.

I could not write a Vim function or a shell command that would make this observation precisely measureable. I guess whenever I want to wrap this operation into a Vim or shell code, Vim realizes that it doesn't actually need to redraw the screen because there it is executing a script.

For normal modules of course (a few hundred lines), this overhead is much smaller.

I would still enable the Markdown support by default, but I understand the concern, and I'm a bit less sure than before. Maybe ask this on Erlang Forums as well.

jchristgit committed:
Add support for triple-quoted strings & docstrings and nominals

Erlang recently added the `-moduledoc` attribute as well as triple
quoted strings and the `~` prefix for binary strings, see [1].

Erlang also added nominal types. See EEP-69 [2].

[1]: https://www.erlang.org/doc/system/documentation.html
[2]: https://www.erlang.org/eeps/eep-0069

Co-authored-by: Csaba Hoch <csaba.hoch@gmail.com>

hcs42 commented:
Great summary in the Erlang Forums post, thanks!

I posted the Erlang Forums link to the `editors` channel on the Erlang Slack. ( https://erlanger.slack.com. I think one can get an invite here: https://erlef.org/slack-invite/erlanger)

hcs42 commented:
*   The poll is going well. I suggest waiting a few more days and then updating the PR with the result (which would be `g:erlang_use_markdown_for_docs = 0` according to the current responses).

*   I found something interesting:

    ```
    -doc """
    This is Markdown text. *italic*, **bold**, `inline code`.

    This is a second paragraph.

    *   First bullet point.

        This is a second paragraph inside the point, but markdown.vim thinks that it
        is a code block. This would be less annoying if at least the highlighting
        color were the same for normal text and code block.

    *   Second bullet point.
    """.
    ```

    Nothing to do here, but this could make someone prefer `g:erlang_docstring_default_highlight = ''` instead of `g:erlang_docstring_default_highlight = 'Comment'`.

    Screenshot:

    ![image](https://github.com/user-attachments/assets/e73b9f84-cab2-4bbd-b9d9-542dc65ab243)

hcs42 commented:
Thank you again, great work!
